src/sentry/static/sentry/app/components/events/interfaces/openInButton.jsx

```diff
@@ -0,0 +1,149 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import SentryTypes from 'app/sentryTypes';
+import Button from 'app/components/button';
+import styled from 'react-emotion';
+
+
+import withApi from 'app/utils/withApi';
+import withLatestContext from 'app/utils/withLatestContext';
+
+class OpenInButton extends React.Component {
+  static propTypes = {
+    api: PropTypes.object,
+    organization: SentryTypes.Organization,
+    project: SentryTypes.Project,
+    lineNo: PropTypes.number,
+    filename: PropTypes.string,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      loading: false,
+      error: false,
+      components: [],
+      sentryApps: [],
+      installs: [],
+    };
+  }
+
+  componentWillMount() {
+    this.fetchIssueLinkComponents();
+  }
+
+  fetchIssueLinkComponents() {
+    const {api, organization, project} = this.props;
+    api
+      .requestPromise(
+        `/organizations/${organization.slug}/sentry-app-components/?filter=stracktrace-link&projectId=${project.id}`
+      )
+      .then(data => {
+        if (data.length) {
+          this.fetchInstallations();
+          this.setState({components: data});
+        }
+      })
+      .catch(error => {
+        return;
+      });
+  }
+
+  fetchInstallations() {
+    const {api, organization} = this.props;
+    api
+      .requestPromise(`/organizations/${organization.slug}/sentry-app-installations/`)
+      .then(data => {
+        if (data.length) {
+          this.fetchSentryApps();
+          this.setState({installs: data});
+        }
+      })
+      .catch(error => {
+        return;
+      });
+  }
+
+  fetchSentryApps() {
+    const {api, organization} = this.props;
+    api
+      .requestPromise(`/organizations/${organization.slug}/sentry-apps/`)
```

Comment:
I'm counting 3 API calls to get sentry-app data. Should we have a 'kitchen sink' API to push the complexity into the server and allow the client to be simpler/more efficient with network?

Reply:
My general opinion is that we already design the API to one specific consumer (this React app) too much. I'd like us to move away from that and towards a world where the API is an interface to the primitives of what makes up Sentry, the system.

Consumers can do whatever makes the most sense for their specific data access/format needs (this could mean the whole "back end for front end" idea or a data layer within the app or something else) in the future.

In any case, I guess I'm saying that I'd personally rather have more small API calls over an endpoint that combines three different resources/concepts.

Also, fwiw, I think using the Reflux Store stuff will help a bit here. We won't need to make these requests over and over since a lot of the data is relatively static and only needs to be loaded once when the Issue page is loaded. I have some work for this in another PR and we can refactor this once it's landed.

Reply:
Ok. My concern was over the additional latency and increased chance for something to fail. I agree that we sometimes build APIs that are more tightly coupled to the react client than they may need to be. We can always come back and add a 'porcelain' API later :smile:

Reply:
I've actually used what matte added and now does some work in the `Preparer` so that we don't need all of those calls 😸